include/graphblas/reference/compressed_storage.hpp

```diff
@@ -30,6 +30,84 @@ namespace grb {
 
 	namespace internal {
 
+		template< typename D, typename IND, typename SIZE >
+		class Compressed_Storage;
+
+		namespace {
+
+			/**
+			 * Copies <tt>row_index<tt> and <tt>col_index<tt> from a
+			 * given Compressed_Storage to another.
+			 *
+			 * Performs no safety checking. Performs no (re-)allocations.
+			 *
+			 * @param[out] output The container to copy the coordinates to.
+			 * @param[in] input   The container to copy the coordinates from.
+			 * @param[in] nz      The number of nonzeroes in the \a other container.
+			 * @param[in] m       The index dimension of the \a other container.
+			 * @param[in] k       The start position to copy from (inclusive).
+			 * @param[in] end     The end position to copy to (exclusive).
+			 *
+			 * The copy range is 2nz + m + 1, i.e.,
+			 *   -# 0 <= start <  2nz + m + 1
+			 *   -# 0 <  end   <= 2nz + m + 1
+			 *
+			 * Concurrent calls to this function are allowed iff they consist of
+			 * disjoint ranges \a start and \a end. The copy is guaranteed to be
+			 * complete if the union of ranges spans 0 to 2nz + m + 1.
+			 */
+			template<
+				typename OutputType, typename OutputIND, typename OutputSIZE,
+				typename InputType, typename InputIND, typename InputSIZE
+			>
+			static inline void copyCoordinatesFrom(
+				Compressed_Storage< OutputType, OutputIND, OutputSIZE > &output,
+				const Compressed_Storage< InputType, InputIND, InputSIZE > &input,
+				const size_t nz, const size_t m,
+				size_t k, size_t end
+			) {
+				static_assert( std::is_convertible< InputIND, OutputIND >::value,
+					"InputIND must be convertible to OutputIND"
+				);
+				static_assert( std::is_convertible< InputSIZE, OutputSIZE >::value,
+					"InputSIZE must be convertible to OutputSIZE"
+				);
+
+				if( k < nz ) {
+					const size_t loop_end = std::min( nz, end );
+					assert( k <= loop_end );
+					std::copy_n(
+						input.row_index + k,
+						loop_end - k,
+						output.row_index + k
+					);
+					k = 0;
+				} else {
+					assert( k >= nz );
+					k -= nz;
+				}
+				if( end <= nz ) {
+					return;
+				}
+				end -= nz;
+				if( k < m + 1 ) {
+					const size_t loop_end = std::min( m + 1, end );
+					assert( k <= loop_end );
+					std::copy_n(
+						input.col_start + k,
+						loop_end - k,
+						output.col_start + k
+					);
+	#ifndef NDEBUG
+					for( size_t chk = k; chk < loop_end - 1; ++chk ) {
+						assert( input.col_start[ chk ] <= input.col_start[ chk + 1 ] );
+						assert( output.col_start[ chk ] <= output.col_start[ chk + 1 ] );
+					}
+	#endif
```

Comment:
Same comment as above:
> Well since this whole logic is called within an already opened OMP region I am not sure it's a good idea. But I can switch to an `simd` for-loop